libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/WaymarkOverlay.kt

```diff
@@ -0,0 +1,159 @@
+            0 -> pushMarker(point, "start", greenWisp)
+            1 -> pushMarker(point, "finish", redWisp)
+            else -> {
+                val prevFinished = finish()
+                popMarker()
+                pushMarker(prevFinished, "waypoint", orangeWisp)
+                pushMarker(point, "finish", redWisp)
+            }
+        }
+    }
+
+    fun removeWaypoint() {
+        when (waymarkersCount()) {
+            0 -> { }
+            1, 2 -> popMarker()
+            else -> {
+                popMarker()
+                val prevFinished = finish()
+                popMarker()
+                pushMarker(prevFinished, "finish", redWisp)
+            }
+        }
+    }
+
+    private fun pushMarker(point: IGeoPoint, label: String, icon: Drawable?) {
+        waymarkers.add(makeMarker(point, label, icon))
+    }
+
+    private fun popMarker() {
+        waymarkers.removeAt(waymarkers.lastIndex)
+    }
+
+    private fun makeMarker(point: IGeoPoint, label: String, icon: Drawable?): OverlayItem {
+        return OverlayItem(label, label, GeoPoint(point.latitude, point.longitude)).apply {
+            setMarker(icon)
+            markerHotspot = OverlayItem.HotspotPlace.BOTTOM_CENTER
+        }
+    }
+
+    ////////////////////////////////////////////
+    override fun draw(canvas: Canvas, mapView: MapView, shadow: Boolean) {
+        val projection = mapView.projection
+
+        waymarkers.forEach { wp -> drawMarker(canvas, projection, wp) }
+    }
+
+    private fun drawMarker(canvas: Canvas,
+                           projection: IProjection,
+                           marker: OverlayItem) {
+        projection.toPixels(marker.point, screenPos)
+
+        val transform = mapView.matrix
+        val transformValues = FloatArray(9)
+        transform.getValues(transformValues)
+
+        val thingToDraw = marker.drawable as BitmapDrawable
+        val halfWidth = thingToDraw.intrinsicWidth / 2
+        val halfHeight = thingToDraw.intrinsicHeight / 2
+        bitmapTransform.apply {
+            setTranslate((-halfWidth).toFloat(), (-halfHeight).toFloat())
+            postScale(1 / transformValues[Matrix.MSCALE_X], 1 / transformValues[Matrix.MSCALE_Y])
+            postTranslate(screenPos.x.toFloat(), screenPos.y.toFloat())
+        }
+        canvas.drawBitmap(thingToDraw.bitmap, bitmapTransform, bitmapPaint)
+    }
+
+    ////////////////////////////////////
+    private fun setWaypoints(waypoints: Waypoints) {
+        resetWaypoints()
+
+        waypoints.forEach { wp -> addWaypoint(wp) }
+    }
+
+    private fun resetWaypoints() {
+        waymarkers.clear()
+    }
+
+    ////////////////////////////////////
+    override fun onResume(prefs: SharedPreferences) {
+        Route.registerListener(this)
+    }
+
+    override fun onPause(prefs: Editor) {
+        Route.unregisterListener(this)
+    }
+
+    override fun onNewJourney(journey: Journey, waypoints: Waypoints) {
+        setWaypoints(waypoints)
```

Comment:
I wonder whether it would be clearer just to inline:
```kotlin
    waymarkers.clear()
    waypoints.forEach { wp -> addWaypoint(wp) }
```

and similarly just the first of those lines in the `onResetJourney()` callback below.

I can't see any other uses of the `setWaypoints()` and `resetWaypoints()` methods.

Reply:
I did think about that, but I decided the names were helpful.